### ISSUE TITLE
Callbacks called twice

### ISSUE THREAD

**Issue by thosta-Witt-Gruppe** [downshift-js/downshift#1447]
- `downshift` version: Tested with `6.1.12`, `7.0.1` and d62ba91e3fd5f2fc3b0fa33f7497e5f78fa3c572
- `node` version: `16.18.1`
- `pnpm` version: `7.15.0`
- `react` version: `18.2.0`

**What you did**:

We are using the `useSelect` hook to display a custom dropdown on larger viewpoints and a modal on smaller viewports.
We recently upgraded to React 18.

**What happened**:

Under certain conditions, the `onSelectedItemChange` callback is called twice when selecting an item.

**Reproduction repository**:

Unfortunately, I am unable to provide a reproduction repo at this point.

**Problem description**:

The problem started to appear when we updated to React 18.
We did some debugging and found that the [`dispatch` function](https://github.com/downshift-js/downshift/blob/7540e3d925ea9eeddf9fe3780b5b4117563e1b1c/src/hooks/utils.js#L176) of downshift's reducer is only called once, however, the actual [reducer function](https://github.com/downshift-js/downshift/blob/7540e3d925ea9eeddf9fe3780b5b4117563e1b1c/src/hooks/useSelect/reducer.js#L8) is called twice under certain circumstances.

According to the [React Docs](https://beta.reactjs.org/apis/react/useReducer#my-reducer-or-initializer-function-runs-twice) and [this issue](https://github.com/facebook/react/issues/16295), reducers should be written in such a way, that them being called twice has no observable effects for the rest of the application.

However, in downshift's case, the state is updated twice with two distinct objects, each containing the exact same information.
This leads to the [useEffect](https://github.com/downshift-js/downshift/blob/7540e3d925ea9eeddf9fe3780b5b4117563e1b1c/src/hooks/utils.js#L181) being called twice and thus [triggering the callback](https://github.com/downshift-js/downshift/blob/7540e3d925ea9eeddf9fe3780b5b4117563e1b1c/src/hooks/utils.js#L183) twice.

**Suggested solution**:

I don't think that we should try to fix the reducer being called twice, due to the lack of a reproduction repo and the fact that React tells developers to expect such behaviour.

Instead, I'd suggest to modify the reducer in such a way, that it being executed twice won't have any observable side effects.
A possible solution might be to replace the following identity check with a deepEqual:
https://github.com/downshift-js/downshift/blob/7540e3d925ea9eeddf9fe3780b5b4117563e1b1c/src/hooks/utils.js#L182
I.e.:
```js
    if (action && prevStateRef.current && !deepEqual(prevStateRef.current, state)) {
```
with `deepEqual` being something like this: https://stackoverflow.com/a/32922084

However, I am not sure if this approach has any unintended consequences.

**Comment by silviuaavram** [comment]
Hi! Thanks for the feedback! It would have been useful to have a repro scenario, so I can figure out what's happening.

**Comment by Fruitseller** [comment]
Hello, we have an issue which is a combination of this one and https://github.com/downshift-js/downshift/issues/1384. We were unable to create a reproduction for that, but have a repo which shows that `onSelectedItemChange` is called twice. Hope the [repo](https://github.com/Fruitseller/downshift-bug-repro) helps to resolve this issue.

**Comment by trainoasis** [comment]
We had the same issue on 7.0.5 & it remains on 7.2.0.

**Comment by nilscox** [comment]
I had a similar behavior when rendering checkboxes in the menu items, because clicking the label triggers the `click` event twice on the menu item element. Here is [a reproduction of the problem](https://codesandbox.io/p/sandbox/wizardly-pare-hprrq9?file=%2Fsrc%2FApp.tsx%3A6%2C1), replacing the `<label />` with anything else solves the issue.

While there is no issue with downshift, I thought it might help some people that may come across this issue.

**Comment by silviuaavram** [comment]
https://codesandbox.io/s/quiet-bash-nyfsqq?file=/src/hooks/useSelect/basic-usage.js

I copied the example from @Fruitseller and I don't get the callback called twice.

Can we get more info so we can dig further? Thanks!

**Comment by klnwsk** [comment]
@silviuaavram I'm also suffering from this issue. @Fruitseller example reproduces the error callback being called twice but only when you use devtools with mobile option because then 2 events happen on each click. One mouse click event and touch event maybe they collide or something, on real mobile device it's the same, callback called twice.

**Comment by okonet** [comment]
@silviuaavram it's reproducible when `useMultipleSelection` is used: https://codesandbox.io/s/musing-lamport-spp6zn?file=/src/hooks/useMultipleSelection/select.js

**Comment by silviuaavram** [comment]
Everyone, try `downshift@8.2.4-alpha.0`. If this solves the issue and does not break anything, we will release it as normal 8.2.4. Thank you!

**Comment by FdelMazo** [comment]
@silviuaavram, beware that the package was released with a typo! it's `downshift@8.2.4-alhpa.0`.

(couldn't get a chance to test it yet)

**Comment by silviuaavram** [comment]
Oups :) you can still test it though and let me know if it works. Thanks!

**Comment by silviuaavram** [comment]
Actually just go for 8.2.4-alpha.1

**Comment by markossankey** [comment]
> Actually just go for 8.2.4-alpha.2

I tried this version, and still seeing the `onSelectedItemChange` being called twice 😬

**Comment by silviuaavram** [comment]
8.2.4 includes the state deep compare. we also had an issue until right now on mobile with controlled selectedItem via formik which triggered an infinite loop on item select, and it got fixed by the deep state compare + formik update.

If there are still issues with these callback changes, I don't think they are downshift related.

**Comment by danichim** [comment]
This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.
![image](https://github.com/downshift-js/downshift/assets/14906788/85ffb9b0-a885-4b4f-ab4e-43fba453c7cd)

**Comment by silviuaavram** [comment]
Hey everyone! I think I got it.

https://github.com/downshift-js/downshift/pull/1571

I also released the 8.3.2-alpha.0 version if you'd like to test it out. Please also check the details in the PR, see if they make sense, and let me know what you think!

**Comment by FdelMazo** [comment]
@silviuaavram can confirm this fixed my issue! Thanks!!
